File: platform/Android/sweeteditor/src/main/java/com/qiplat/sweeteditor/EditorSettings.java
package com.qiplat.sweeteditor;

import android.graphics.Typeface;

import androidx.annotation.NonNull;
import androidx.annotation.MainThread;

import com.qiplat.sweeteditor.core.EditorCore;
import com.qiplat.sweeteditor.core.foundation.AutoIndentMode;
import com.qiplat.sweeteditor.core.foundation.CurrentLineRenderMode;
import com.qiplat.sweeteditor.core.foundation.FoldArrowMode;
import com.qiplat.sweeteditor.core.foundation.WrapMode;

/**
 * Centralized configuration for {@link SweetEditor}.
 * <p>
 * Obtain via {@link SweetEditor#getSettings()}. All setters take effect immediately.
 * Public APIs on this class are main-thread only.
 */
@MainThread
public class EditorSettings {

    private final SweetEditor mEditor;

    private float mTextSize = 36f;
    private Typeface mTypeface = Typeface.create(Typeface.MONOSPACE, Typeface.NORMAL);
    private float mScale = 1.0f;
    private FoldArrowMode mFoldArrowMode = FoldArrowMode.ALWAYS;
    private WrapMode mWrapMode = WrapMode.NONE;
    private float mLineSpacingAdd = 0f;
    private float mLineSpacingMult = 1.0f;
    private float mContentStartPadding = 0f;
    private boolean mShowSplitLine = true;
    private boolean mGutterSticky = false;
    private boolean mGutterVisible = true;
    private CurrentLineRenderMode mCurrentLineRenderMode = CurrentLineRenderMode.BACKGROUND;
    private AutoIndentMode mAutoIndentMode = AutoIndentMode.KEEP_INDENT;
    private boolean mBackspaceUnindent = true;
    private boolean mReadOnly = false;
    private int mMaxGutterIcons = 0;
    private long mDecorationScrollRefreshMinIntervalMs = 16L;
    private float mDecorationOverscanViewportMultiplier = 1.5f;
    private boolean mCursorAnimationEnabled = true;

    EditorSettings(@NonNull SweetEditor editor) {
        mEditor = editor;
    }

    public void setEditorTextSize(float textSize) {
        mTextSize = textSize;
        mEditor.applyTextSize(textSize);
    }

    public float getEditorTextSize() {
        return mTextSize;
    }

    public void setTypeface(@NonNull Typeface typeface) {
        mTypeface = typeface;
        mEditor.applyTypeface(typeface);
    }

    @NonNull
    public Typeface getTypeface() {
        return mTypeface;
    }

    public void setScale(float scale) {
        mScale = scale;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setScale(scale);
        mEditor.dispatchEditorActionResult(result);
    }

    public float getScale() {
        return mScale;
    }

    public void setFoldArrowMode(@NonNull FoldArrowMode mode) {
        mFoldArrowMode = mode;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setFoldArrowMode(mode.value);
        mEditor.dispatchEditorActionResult(result);
    }

    @NonNull
    public FoldArrowMode getFoldArrowMode() {
        return mFoldArrowMode;
    }

    public void setWrapMode(@NonNull WrapMode mode) {
        mWrapMode = mode;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setWrapMode(mode.value);
        mEditor.dispatchEditorActionResult(result);
    }

    @NonNull
    public WrapMode getWrapMode() {
        return mWrapMode;
    }

    public void setLineSpacing(float add, float mult) {
        mLineSpacingAdd = add;
        mLineSpacingMult = mult;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setLineSpacing(add, mult);
        mEditor.dispatchEditorActionResult(result);
    }

    public float getLineSpacingAdd() {
        return mLineSpacingAdd;
    }

    public float getLineSpacingMult() {
        return mLineSpacingMult;
    }

    public void setContentStartPadding(float padding) {
        mContentStartPadding = Math.max(0f, padding);
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setContentStartPadding(mContentStartPadding);
        mEditor.dispatchEditorActionResult(result);
    }

    public float getContentStartPadding() {
        return mContentStartPadding;
    }

    public void setShowSplitLine(boolean show) {
        mShowSplitLine = show;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setShowSplitLine(show);
        mEditor.dispatchEditorActionResult(result);
    }

    public boolean isShowSplitLine() {
        return mShowSplitLine;
    }

    public void setGutterSticky(boolean sticky) {
        mGutterSticky = sticky;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setGutterSticky(sticky);
        mEditor.dispatchEditorActionResult(result);
    }

    public boolean isGutterSticky() {
        return mGutterSticky;
    }

    public void setGutterVisible(boolean visible) {
        mGutterVisible = visible;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setGutterVisible(visible);
        mEditor.dispatchEditorActionResult(result);
    }

    public boolean isGutterVisible() {
        return mGutterVisible;
    }

    public void setCurrentLineRenderMode(@NonNull CurrentLineRenderMode mode) {
        mCurrentLineRenderMode = mode;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setCurrentLineRenderMode(mode.value);
        mEditor.dispatchEditorActionResult(result);
    }

    @NonNull
    public CurrentLineRenderMode getCurrentLineRenderMode() {
        return mCurrentLineRenderMode;
    }

    public void setAutoIndentMode(@NonNull AutoIndentMode mode) {
        mAutoIndentMode = mode;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setAutoIndentMode(mode.value);
        mEditor.dispatchEditorActionResult(result);
    }

    @NonNull
    public AutoIndentMode getAutoIndentMode() {
        return mAutoIndentMode;
    }

    public void setBackspaceUnindent(boolean enabled) {
        mBackspaceUnindent = enabled;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setBackspaceUnindent(enabled);
        mEditor.dispatchEditorActionResult(result);
    }

    public boolean isBackspaceUnindent() {
        return mBackspaceUnindent;
    }

    public void setReadOnly(boolean readOnly) {
        mReadOnly = readOnly;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setReadOnly(readOnly);
        mEditor.dispatchEditorActionResult(result);
    }

    public boolean isReadOnly() {
        return mReadOnly;
    }

    public void setMaxGutterIcons(int count) {
        mMaxGutterIcons = count;
        EditorCore.EditorActionResult result = mEditor.getEditorCore().setMaxGutterIcons(count);
        mEditor.dispatchEditorActionResult(result);
    }

    public int getMaxGutterIcons() {
        return mMaxGutterIcons;
    }

    public void setDecorationScrollRefreshMinIntervalMs(long intervalMs) {
        mDecorationScrollRefreshMinIntervalMs = Math.max(0L, intervalMs);
        mEditor.requestDecorationRefresh();
    }

    public long getDecorationScrollRefreshMinIntervalMs() {
        return mDecorationScrollRefreshMinIntervalMs;
    }

    public void setDecorationOverscanViewportMultiplier(float multiplier) {
        mDecorationOverscanViewportMultiplier = Math.max(0f, multiplier);
        mEditor.requestDecorationRefresh();
    }

    public float getDecorationOverscanViewportMultiplier() {
        return mDecorationOverscanViewportMultiplier;
    }

    public void setCursorAnimationEnabled(boolean enabled) {
        mCursorAnimationEnabled = enabled;
        mEditor.requestCursorAnimationRefresh();
    }

    public boolean isCursorAnimationEnabled() {
        return mCursorAnimationEnabled;
    }

}

File: platform/Apple/SweetEditor-Shared/EditorSettings.swift
import Foundation

package enum EditorSettingChange {
    case font(size: Float, typeface: String)
    case scale(Float)
    case foldArrowMode(FoldArrowMode)
    case wrapMode(WrapMode)
    case renderWhitespace(WhitespaceRenderMode)
    case renderLineBreaks(Bool)
    case lineSpacing(add: Float, mult: Float)
    case contentStartPadding(Float)
    case showSplitLine(Bool)
    case gutterSticky(Bool)
    case gutterVisible(Bool)
    case currentLineRenderMode(CurrentLineRenderMode)
    case autoIndentMode(AutoIndentMode)
    case backspaceUnindent(Bool)
    case readOnly(Bool)
    case compositionEnabled(Bool)
    case maxGutterIcons(UInt32)
    case decorationOverscanViewportMultiplier
}

/// Centralized runtime editor configuration shared by Apple platform bindings.
///
/// This type mirrors the Android-side `EditorSettings` design: runtime behavior knobs
/// such as scale, wrapping, read-only mode, line spacing, split-line visibility, and
/// gutter/icon limits live here and take effect immediately when mutated.
///
/// Keep theme, language configuration, highlights, diagnostics, and provider-style
/// integrations outside this type. Those are separate editor concepts and should stay
/// on their dedicated APIs.
public final class EditorSettings {
    private let onChange: (EditorSettingChange) -> Void

    public private(set) var editorTextSize: Float
    public private(set) var typeface: String = "Menlo"
    public private(set) var scale: Float = 1.0
    public private(set) var foldArrowMode: FoldArrowMode = .always
    public private(set) var wrapMode: WrapMode = .none
    public private(set) var renderWhitespace: WhitespaceRenderMode = .none
    public private(set) var renderLineBreaks = false
    public private(set) var lineSpacingAdd: Float = 0.0
    public private(set) var lineSpacingMult: Float = 1.2
    public private(set) var contentStartPadding: Float = 3.0
    public private(set) var showSplitLine = true
    public private(set) var gutterSticky: Bool
    public private(set) var gutterVisible = true
    public private(set) var currentLineRenderMode: CurrentLineRenderMode = .background
    public private(set) var autoIndentMode: AutoIndentMode = .keepIndent
    public private(set) var backspaceUnindent = true
    public private(set) var readOnly = false
    public private(set) var compositionEnabled = true
    public private(set) var maxGutterIcons: UInt32 = 0
    public private(set) var decorationScrollRefreshMinIntervalMs: Int64 = 16
    public private(set) var decorationOverscanViewportMultiplier: Float = 1.5

    package init(editorTextSize: Float,
                 gutterSticky: Bool,
                 onChange: @escaping (EditorSettingChange) -> Void) {
        self.editorTextSize = editorTextSize
        self.gutterSticky = gutterSticky
        self.onChange = onChange
    }

    /// Updates the editor text size and applies the change immediately.
    public func setEditorTextSize(_ textSize: Float) {
        editorTextSize = max(1, textSize)
        apply(.font(size: editorTextSize, typeface: typeface))
    }

    /// Updates the editor typeface and applies the change immediately.
    public func setTypeface(_ typeface: String) {
        self.typeface = typeface
        apply(.font(size: editorTextSize, typeface: typeface))
    }

    /// Updates editor scale and applies the change immediately.
    public func setScale(_ scale: Float) {
        self.scale = scale
        apply(.scale(scale))
    }

    /// Updates fold-arrow rendering mode and applies the change immediately.
    public func setFoldArrowMode(_ mode: FoldArrowMode) {
        foldArrowMode = mode
        apply(.foldArrowMode(mode))
    }

    /// Updates wrapping mode and applies the change immediately.
    public func setWrapMode(_ mode: WrapMode) {
        wrapMode = mode
        apply(.wrapMode(mode))
    }

    /// Updates whitespace marker rendering mode and applies the change immediately.
    public func setRenderWhitespace(_ mode: WhitespaceRenderMode) {
        renderWhitespace = mode
        apply(.renderWhitespace(mode))
    }

    /// Shows or hides line-break markers and applies the change immediately.
    public func setRenderLineBreaks(_ enabled: Bool) {
        renderLineBreaks = enabled
        apply(.renderLineBreaks(enabled))
    }

    /// Updates line spacing and applies the change immediately.
    public func setLineSpacing(add: Float, mult: Float) {
        lineSpacingAdd = add
        lineSpacingMult = mult
        apply(.lineSpacing(add: add, mult: mult))
    }

    /// Updates leading content padding and applies the change immediately.
    public func setContentStartPadding(_ padding: Float) {
        contentStartPadding = max(0, padding)
        apply(.contentStartPadding(contentStartPadding))
    }

    /// Shows or hides the split line and applies the change immediately.
    public func setShowSplitLine(_ show: Bool) {
        showSplitLine = show
        apply(.showSplitLine(show))
    }

    /// Controls whether the gutter remains fixed during horizontal scrolling.
    public func setGutterSticky(_ sticky: Bool) {
        gutterSticky = sticky
        apply(.gutterSticky(sticky))
    }

    /// Shows or hides the gutter area.
    public func setGutterVisible(_ visible: Bool) {
        gutterVisible = visible
        apply(.gutterVisible(visible))
    }

    /// Updates current-line rendering mode and applies the change immediately.
    public func setCurrentLineRenderMode(_ mode: CurrentLineRenderMode) {
        currentLineRenderMode = mode
        apply(.currentLineRenderMode(mode))
    }

    /// Updates auto-indent behavior and applies the change immediately.
    public func setAutoIndentMode(_ mode: AutoIndentMode) {
        autoIndentMode = mode
        apply(.autoIndentMode(mode))
    }

    /// Updates backspace unindent behavior and applies the change immediately.
    public func setBackspaceUnindent(_ enabled: Bool) {
        backspaceUnindent = enabled
        apply(.backspaceUnindent(enabled))
    }

    /// Updates read-only mode and applies the change immediately.
    public func setReadOnly(_ readOnly: Bool) {
        self.readOnly = readOnly
        apply(.readOnly(readOnly))
    }

    public func setCompositionEnabled(_ enabled: Bool) {
        compositionEnabled = enabled
        apply(.compositionEnabled(enabled))
    }

    /// Updates the maximum visible gutter icon count and applies the change immediately.
    public func setMaxGutterIcons(_ count: UInt32) {
        maxGutterIcons = count
        apply(.maxGutterIcons(count))
    }

    public func setDecorationScrollRefreshMinIntervalMs(_ intervalMs: Int64) {
        decorationScrollRefreshMinIntervalMs = max(0, intervalMs)
    }

    public func setDecorationOverscanViewportMultiplier(_ multiplier: Float) {
        decorationOverscanViewportMultiplier = max(0, multiplier)
        apply(.decorationOverscanViewportMultiplier)
    }

    package func applyAll() {
        let changes: [EditorSettingChange] = [
            .font(size: editorTextSize, typeface: typeface),
            .scale(scale),
            .foldArrowMode(foldArrowMode),
            .wrapMode(wrapMode),
            .renderWhitespace(renderWhitespace),
            .renderLineBreaks(renderLineBreaks),
            .lineSpacing(add: lineSpacingAdd, mult: lineSpacingMult),
            .contentStartPadding(contentStartPadding),
            .showSplitLine(showSplitLine),
            .gutterSticky(gutterSticky),
            .gutterVisible(gutterVisible),
            .currentLineRenderMode(currentLineRenderMode),
            .autoIndentMode(autoIndentMode),
            .backspaceUnindent(backspaceUnindent),
            .readOnly(readOnly),
            .compositionEnabled(compositionEnabled),
            .maxGutterIcons(maxGutterIcons),
        ]
        for change in changes {
            apply(change)
        }
    }

    private func apply(_ change: EditorSettingChange) {
        onChange(change)
    }
}
